File: core/src/main/scala/tripaint/coords/StorageCoords.scala
package tripaint.coords

case class StorageCoords(value: Long) extends AnyVal {
  inline def x: Int = (value >> 32).toInt
  inline def y: Int = value.toInt
}

object StorageCoords {
  inline def apply(x: Int, y: Int): StorageCoords = {
    require(x >= 0)
    require(y >= 0)
    new StorageCoords(x.toLong << 32 | y)
  }
}

File: core/src/main/kotlin/tripaint/coords/StorageCoords.kt
package tripaint.coords

data class StorageCoords(val value: Long) {
    val x: Int
        get() = (value shr 32).toInt()
    val y: Int
        get() = value.toInt()

    companion object {
        fun from(x: Int, y: Int): StorageCoords {
            require(x >= 0)
            require(y >= 0)
            return StorageCoords((x.toLong() shl 32) or y.toLong())
        }
    }
}
